plugins/jobtype/src/azkaban/jobtype/HadoopSecureSparkWrapper.java

```diff
@@ -62,47 +86,142 @@ public static void main(final String[] args) throws Exception {
+    boolean dynamicAllocEnabled = dynamicAllocProp != null && dynamicAllocProp.equals(Boolean.TRUE.toString());
+    if (dynamicAllocEnabled) {
+      for (int i = 0; i < argArray.length; i++) {
+        if (argArray[i] == null) continue;
+
+        // If user specifies num of executors, or if user tries to disable dynamic allocation for his application
+        // by setting some conf params to falase, we need to ignore these settings to enforce the application 
+        // uses dynamic allocation.
+        if (argArray[i].equals(SparkJobArg.NUM_EXECUTORS.sparkParamName) // --num-executors
+          || (
+            argArray[i].equals(SparkJobArg.SPARK_CONF_PREFIX.sparkParamName) // --conf
+            && (argArray[i+1].startsWith(SPARK_CONF_SHUFFLE_SERVICE_ENABLED) // spark.shuffle.service.enabled=
+              || argArray[i+1].startsWith(SPARK_CONF_DYNAMIC_ALLOC_ENABLED)) // spark.dynamicAllocation.enabled
+        )) {
+
+          logger.info("Azbakan enforces dynamic resource allocation. Ignore user param: "
+            + argArray[i] + " " + argArray[i+1]);
+          argArray[i] = null;
+          argArray[++i] = null;
+        }
+      }
+    }
+  }
+
+  private static void handleNodeLabeling(String[] argArray) {
+    // HadoopSparkJob will set env var on this process if we enable node labeling policy for spark jobtype.
+    // We also detect the yarn cluster settings has enable node labeling
+    // We enforce node labeling policy by ignoring user queue params.
+    Configuration conf = new Configuration();
+    boolean nodeLabelingYarn = conf.getBoolean(YARN_CONF_NODE_LABELING_ENABLED, false);
+    String nodeLabelingProp = System.getenv(HadoopSparkJob.SPARK_NODE_LABELING_ENV_VAR);
+    boolean nodeLabelingPolicy = nodeLabelingProp != null && nodeLabelingProp.equals(Boolean.TRUE.toString());
+
+    if (nodeLabelingYarn && nodeLabelingPolicy) {
+      for (int i = 0; i < argArray.length; i++) {
+        if (argArray[i] == null) continue;
+        // If yarn cluster enables node labeling, applications should be submitted to a default
+        // queue by a default conf(spark.yarn.queue) in spark-defaults.conf 
+        // We should ignore user-specified queue param to enforece the node labeling
+        // (--queue test or --conf spark.yarn.queue=test)
+        if ((argArray[i].equals(SparkJobArg.SPARK_CONF_PREFIX.sparkParamName) &&
+             argArray[i+1].startsWith(SPARK_CONF_QUEUE))
+            || (argArray[i].equals(SparkJobArg.QUEUE.sparkParamName))) {
+
+          logger.info("Azbakan enforces node labeling. Ignore user param: "
+            + argArray[i] + " " + argArray[i+1]);
+          argArray[i] = null;
+          argArray[++i] = null;
+        }
+      }
+    }
   }
 
+  private static String[] removeNullsFromArgArray(String[] argArray) {
+    List<String> argList = new ArrayList(Arrays.asList(argArray));
+    argList.removeAll(Collections.singleton(null));
+    return argList.toArray(new String[0]);
```

Comment:
Why not "return argList.toArray(new String[argList.size()]);" so the String array with the appropriate length is preallocated?